src/solve.jl

```diff
@@ -17,71 +17,493 @@ function DiffEqBase.__solve(prob::BVProblem, alg::Shooting; kwargs...)
+    #exports: mesh_new, Nsub_star, info
+
+    #TODO: Need users to manually specify, here, we set it as 3000 by default.
+    MxNsub = 3000
+
+    safety_factor = 1.3
+    rho = 1.0 # Set rho=1 means mesh distribution will take place everytime.
+    upper_new_mesh = 4.0
+    lower_new_mesh = 0.5
+    r1 = 0.0
+    r2 = 0.0
+    Nsub_star = 0
+    info = 0
+    p = alg_order(alg)
+    s_hat = zeros(Float64, n)
+    for i in 1:n
+        h = mesh_current[i + 1] - mesh_current[i]
+        norm = abs(defect[i, idamax(defect[i, :])])
+        s_hat[i] = (norm / tol)^(1.0 / (p + 1)) / h
+        if s_hat[i] * h > r1
+            r1 = s_hat[i] * h
+        end
+        r2 = r2 + s_hat[i] * h
+    end
+    r3 = r2 / n
+    n_predict::Int64 = round(Int, (safety_factor * r2) + 1)
+    if abs((n_predict - n) / n) < 0.1
+        n_predict = round(Int, 1.1 * n)
+    end
+
+    if r1 <= rho * r3
+        Nsub_star::Int64 = 2 * n
+        if Nsub_star > MxNsub # Need to determine the too large threshold
+            println("New mesh would be too large")
+            info = -1
         else
-            general_eval_bc_residual!(S)
+            println("Half the current mesh")
+            mesh_new = half_mesh(mesh_current, n)
         end
-        flatten_vector!(resid, S.residual)
-        reorder!(resid)
-        return nothing
+    else
+        Nsub_star = copy(n_predict)
+        if Nsub_star > upper_new_mesh * n
+            Nsub_star = upper_new_mesh * n
+        end
+        if Nsub_star < lower_new_mesh * n
+            Nsub_star = lower_new_mesh * n
+        end
+        if Nsub_star > MxNsub
+            # Mesh redistribution fails
+            println("New mesh would be too large")
+            info = -1
+        else
+            println("Mesh redistributing")
+            mesh_new = redistribute(mesh_current, n, Nsub_star, s_hat)
+        end
+    end
+    return mesh_new, Nsub_star, info
+end
+
+"""
+    redistribute(mesh_current, n, Nsub_star, s_hat)
+
+Generate a new mesh.
+"""
+function redistribute(mesh_current::Vector,
+                      n::Int64,
+                      Nsub_star::Int64,
+                      s_hat::Vector{Float64})
+    mesh_new = zeros(Float64, Nsub_star + 1)
+    sum = 0.0
+    for k in 1:n
+        sum += s_hat[k] * (mesh_current[k + 1] - mesh_current[k])
+    end
+    zeta = sum / Nsub_star
+    k::Int64 = 1
+    i::Int64 = 0
+    mesh_new[1] = mesh_current[1]
+    t = mesh_current[1]
+    integral::Float64 = 0.0
+    while k <= n
+        next_piece = s_hat[k] * (mesh_current[k + 1] - t)
+        if (integral + next_piece) > zeta
+            mesh_new[i + 2] = (zeta - integral) / s_hat[k] + t
+            t = mesh_new[i + 2]
+            i += 1
+            integral = 0
+        else
+            integral += next_piece
+            t = mesh_current[k + 1]
+            k += 1
+        end
+    end
+    mesh_new[Nsub_star + 1] = mesh_current[n + 1]
+    return mesh_new
+end
```

Comment:
Can you describe what this one is doing?

Reply:
The redistribute part generates a new mesh to make sure that on the next iteration, the new solution on the new mesh would have a smaller defect estimation.

Based on the judging quantities we computed before using defect and mesh function values, we compare the ratio $\frac{r1}{r3}$ with $\rho$ to decide whether we halve the mesh or make a redistribute.
```julia
    s_hat = zeros(Float64, n)
    for i in 1:n
        h = mesh_current[i + 1] - mesh_current[i]
        norm = abs(defect[i, idamax(defect[i, :])])
        s_hat[i] = (norm / tol)^(1.0 / (p + 1)) / h
        if s_hat[i] * h > r1
            r1 = s_hat[i] * h
        end
        r2 = r2 + s_hat[i] * h
    end
    r3 = r2 / n
    n_predict::Int64 = round(Int, (safety_factor * r2) + 1)
    if abs((n_predict - n) / n) < 0.1
        n_predict = round(Int, 1.1 * n)
    end
```
 If we need to redistribute(r1/r3>rho), we redistribute a new mesh, I keep some detailed notes on the redistribute process here: https://erikqqy.github.io/2023/06/04/gsoc-1st-week/#Redistribute-mesh